Marlin/src/gcode/feature/L6470/M916-918.cpp

```diff
@@ -439,91 +533,124 @@ void GcodeSuite::M917() {
+ *     optional - will report current value from driver if not specified
+ *
+ * M - value for microsteps (1 - 128) (optional)
  *     optional - will report current value from driver if not specified
  *
  */
 void GcodeSuite::M918() {
 
   DEBUG_ECHOLNPGM("M918");
 
-  char axis_mon[3][3] = { "  ", "  ", "  " };  // List of axes to monitor
-  uint8_t axis_index[3];
+  L64xx_active = true;               //  keep monitor_driver() from stealing status
+
+  char axis_mon[3][3] = { {"  "}, {"  "}, {"  "} };   // list of Axes to be monitored
+  L64XX_axis_t axis_index[3];
   uint16_t axis_status[3];
   uint8_t driver_count = 1;
   float position_max, position_min;
-  feedRate_t final_fr_mm_m;
+  float final_feedrate;
   uint8_t kval_hold;
-  uint8_t ocd_th_val = 0;
-  uint8_t stall_th_val = 0;
+  uint8_t OCD_TH_val = 0;
+  uint8_t STALL_TH_val = 0;
   uint16_t over_current_threshold;
-  constexpr bool over_current_flag = true;
+  constexpr uint8_t over_current_flag = true;
+
+  const L64XX_Marlin::L64XX_shadow_t &sh = L64xxManager.shadow;
 
   uint8_t j;   // general purpose counter
 
-  if (L6470.get_user_input(driver_count, axis_index, axis_mon, position_max, position_min, final_fr_mm_m, kval_hold, over_current_flag, ocd_th_val, stall_th_val, over_current_threshold))
+  if (L64xxManager.get_user_input(driver_count, axis_index, axis_mon, position_max, position_min, final_feedrate, kval_hold, over_current_flag, OCD_TH_val, STALL_TH_val, over_current_threshold))
     return;  // quit if invalid user input
 
+  L64xxManager.get_status(axis_index[0]); // populate shadow array
+
   uint8_t m_steps = parser.byteval('M');
-  LIMIT(m_steps, 0, 128);
-  DEBUG_ECHOLNPAIR("M = ", m_steps);
-
-  int8_t m_bits = -1;
-       if (m_steps > 85) m_bits = 7;  // 128 (no synch output)
-  else if (m_steps > 42) m_bits = 6;  //  64 (no synch output)
-  else if (m_steps > 22) m_bits = 5;  //  32 (no synch output)
-  else if (m_steps > 12) m_bits = 4;  //  16 (no synch output)
-  else if (m_steps >  5) m_bits = 3;  //   8 (no synch output)
-  else if (m_steps >  2) m_bits = 2;  //   4 (no synch output)
-  else if (m_steps == 2) m_bits = 1;  //   2 (no synch output)
-  else if (m_steps == 1) m_bits = 0;  //   1 (no synch output)
-  else if (m_steps == 0) m_bits = 7;  // 128 (no synch output)
-
-  if (m_bits >= 0) {
-    const int micros = _BV(m_bits);
-    if (micros < 100) { DEBUG_CHAR(' '); if (micros < 10) DEBUG_CHAR(' '); }
-    DEBUG_ECHO(micros);
-    DEBUG_ECHOPGM(" uSTEPS");
+
+  if (m_steps != 0) {
+    if (sh.STATUS_AXIS_LAYOUT == 1) LIMIT(m_steps, 1, 16);  // L6474
+    else                            LIMIT(m_steps, 1, 128);
+
+    uint8_t stepVal;
+    for (stepVal = 0; stepVal < 8; stepVal++) {  // convert to L64xx register value
+      if (m_steps == 1) break;
+      m_steps >>= 1;
+    }
+
+    if (sh.STATUS_AXIS_LAYOUT == 1)
+      stepVal |= 0x98 ;  //no sync
+    else
+      stepVal |= (!SYNC_EN) | SYNC_SEL_1 | stepVal;
```

Comment:
Is `!SYNC_EN` correct here? If `SYNC_EN` is a bit mask (or anything non-zero) then `!` will make it always zero. If it is zero then this will always evaluate to 1.

Reply:
I see that a patch came in to the `dev` branch of Arduino-L6470 which changes `!SYNC_EN` to `BUSY_EN` so that may be the correction here as well.